### PR TITLE
St/mempool size check

### PR DESCRIPTION
**Pull request by saratnt**
This PR refers to EVM-647 and EVM-648 JIRA tasks.

### CODE REVIEW

File: qa/sc_evm_mempool_size.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+Check mem pool behaviour when it reached its maximum size.
+
+Configuration:
+    - 1 SC node
+    - 1 MC node
+
+"""
+
+
+class SCEvmMempoolSize(AccountChainSetup):
```
[the remark this answers is not in the document]
Comment (saratnt) [reply qa/sc_evm_mempool_size.py:24]: Already solved in PR: https://github.com/HorizenOfficial/Sidechains-SDK/pull/744

File: sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala [hunk 2]
```diff
@@ -24,9 +24,13 @@ class MempoolMap(
 
   val MaxNonceGap: BigInteger = BigInteger.valueOf(mempoolSettings.maxNonceGap)
   val MaxSlotsPerAccount: Int = mempoolSettings.maxAccountSlots
+  val MaxMemPoolSlots: Int = mempoolSettings.maxMemPoolSlots
+
+//  // All transactions currently in the mempool
+//  private val all: TrieMap[ModifierId, SidechainTypes#SCAT] = TrieMap.empty[ModifierId, SidechainTypes#SCAT]
```
[the remark this answers is not in the document]
Comment (saratnt) [reply sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala:30]: Done

File: sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala [hunk 3]
```diff
@@ -36,25 +40,57 @@ class MempoolMap(
   // Next expected nonce for each account. It is the max nonce of the executable txs plus 1, if any, otherwise it has the same value of the statedb nonce.
   private val nonces: TrieMap[SidechainTypes#SCP, BigInteger] = TrieMap.empty[SidechainTypes#SCP, BigInteger]
 
+  private def findTxWithSameNonce(account: SidechainTypes#SCP, nonce: BigInteger): Option[SidechainTypes#SCAT] = {
+    val txOpt: Option[SidechainTypes#SCAT] = executableTxs.get(account).flatMap(map => map.get(nonce).map(txCache(_)))
```
[the remark this answers is not in the document]
Comment (saratnt) [reply sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala:44]: Done

File: sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala [hunk 4]
```diff
@@ -36,25 +40,57 @@ class MempoolMap(
...
         log.trace(s"Transaction $ethTransaction nonce gap respect state nonce exceeds maximum allowed size: tx nonce ${ethTransaction.getNonce}, " +
           s"state nonce: $stateNonce")
         throw NonceGapTooWideException(ethTransaction.id, ethTransaction.getNonce, stateNonce)
       }
 
+      // Reject transactions in case their account doesn't have enough space for it. If the new tx is replacing an
+      // existing one, the space used by the old one must be taken into account.
+
+      val txToReplace = findTxWithSameNonce(account, ethTransaction.getNonce)
+      val numOfSlotsOfTxToReplace = if (txToReplace.isDefined){
```
[the remark this answers is not in the document]
Comment (saratnt) [reply sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala:75]: Done

File: sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala [hunk 5]
```diff
@@ -80,70 +116,65 @@ class MempoolMap(
...
             val existingTxWithSameNonceId = existingTxWithSameNonceIdOpt.get
-            replaceIfCanPayHigherFee(existingTxWithSameNonceId, ethTransaction, nonExecTxsPerAccount)
+            replaceTransaction(existingTxWithSameNonceId, ethTransaction, nonExecTxsPerAccount)
           } else {
             addNewTransaction(nonExecTxsPerAccount, ethTransaction)
           }
         case ReplaceExecTransaction =>
           // This case means there is already an executable tx with the same nonce in the mem pool
           val executableTxsPerAccount = executableTxs(account)
           val existingTxWithSameNonceId = executableTxsPerAccount(ethTransaction.getNonce)
```
[the remark this answers is not in the document]
Comment (saratnt) [reply sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala:132]: Done

File: sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala [hunk 6]
```diff
@@ -80,70 +116,65 @@ class MempoolMap(
...
-          replaceIfCanPayHigherFee(existingTxWithSameNonceId, ethTransaction, executableTxsPerAccount)
+          replaceTransaction(existingTxWithSameNonceId, ethTransaction, executableTxsPerAccount)
+      }
+
+      //After having added the new tx, check the resulting size of the mempool. If it exceeds the maximum, free some space.
+      if (getMempoolSizeInSlots > MaxMemPoolSlots) {
```
[the remark this answers is not in the document]
Comment (saratnt) [reply sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala:137]: Already done in PR: https://github.com/HorizenOfficial/Sidechains-SDK/pull/744

File: sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala [hunk 7]
```diff
@@ -80,70 +116,65 @@ class MempoolMap(
...
-      log.trace(s"Replacing transaction $existingTxWithSameNonce with $newTx")
-      all.remove(existingTxId)
-      all.put(newTx.id, newTx)
-      mapOfTxsByNonce.put(newTx.getNonce, newTx.id)
-    }
-    else {
-      log.trace(s"Transaction $newTx cannot replace $existingTxWithSameNonce because it is underpriced")
-      throw TransactionReplaceUnderpricedException(newTx.id)
+
+  private[mempool] def freeMempoolSlots(additionalSlotsToFree: Int = 0): Unit = {
```
[the remark this answers is not in the document]
Comment (saratnt) [reply sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala:161]: Actually no, it was removed iin PR: https://github.com/HorizenOfficial/Sidechains-SDK/pull/744

File: sdk/src/main/scala/com/horizen/account/mempool/TxCache.scala [hunk 8]
```diff
@@ -0,0 +1,82 @@
+package com.horizen.account.mempool
+
+import com.horizen.SidechainTypes
+import com.horizen.account.mempool.MempoolMap.txSizeInSlot
+import sparkz.util.ModifierId
+
+import scala.collection.concurrent.TrieMap
+
+class TxCache {
```
[the remark this answers is not in the document]
Comment (saratnt) [reply sdk/src/main/scala/com/horizen/account/mempool/TxCache.scala:9]: Done

File: sdk/src/main/scala/com/horizen/account/mempool/TxCache.scala [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+  // All transactions currently in the mempool
+  private val all: TrieMap[ModifierId, TxMetaInfo] = TrieMap.empty[ModifierId, TxMetaInfo]
+  private var sizeInSlots: Int = 0
+
+  private var oldestTx: Option[TxMetaInfo] = None
+  private var youngestTx: Option[TxMetaInfo] = None
+
+  def add(tx: SidechainTypes#SCAT): Unit = {
+    val txInf0 = TxMetaInfo(tx)
```
[the remark this answers is not in the document]
Comment (saratnt) [reply sdk/src/main/scala/com/horizen/account/mempool/TxCache.scala:18]: Done

File: sdk/src/main/scala/com/horizen/account/mempool/TxCache.scala [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+  def getSizeInSlots: Int = sizeInSlots
+
+  def getOldestTransaction(): Option[SidechainTypes#SCAT] = oldestTx.map(_.tx)
+
+  def getYoungestTransaction(): Option[SidechainTypes#SCAT] = youngestTx.map(_.tx)
+
+
+  case class TxMetaInfo(tx: SidechainTypes#SCAT) {
+    var next: Option[TxMetaInfo] = None
+    var previous: Option[TxMetaInfo] = None
```
[the remark this answers is not in the document]
Comment (saratnt) [reply sdk/src/main/scala/com/horizen/account/mempool/TxCache.scala:78]: Done

File: sdk/src/main/scala/com/horizen/account/mempool/TxCache.scala [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+  def contains(txId: ModifierId): Boolean = all.contains(txId)
+
+  def getSizeInSlots: Int = sizeInSlots
+
+  def getOldestTransaction(): Option[SidechainTypes#SCAT] = oldestTx.map(_.tx)
+
+  def getYoungestTransaction(): Option[SidechainTypes#SCAT] = youngestTx.map(_.tx)
+
+
+  case class TxMetaInfo(tx: SidechainTypes#SCAT) {
```
[the remark this answers is not in the document]
Comment (saratnt) [reply sdk/src/main/scala/com/horizen/account/mempool/TxCache.scala:76]: Done

File: sdk/src/main/scala/com/horizen/account/mempool/TxCache.scala [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+      tmp.get.next = youngestTx
+    }
+  }
+
+  def remove(txId: ModifierId): Option[SidechainTypes#SCAT] = {
+    val txInfoOpt = all.remove(txId)
+    txInfoOpt.map { txInfo =>
+      val prev = txInfo.previous
+      val next = txInfo.next
+      if (prev.isEmpty && next.isEmpty){
```
[the remark this answers is not in the document]
Comment (saratnt) [reply sdk/src/main/scala/com/horizen/account/mempool/TxCache.scala:37]: Done

File: sdk/src/main/scala/com/horizen/account/mempool/TxCache.scala [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+      txInfo.tx
+    }
+
+  }
+
+  def getTransaction(txId: ModifierId): Option[SidechainTypes#SCAT] = all.get(txId).map(_.tx)
+
+  def apply(txId: ModifierId): SidechainTypes#SCAT = all(txId).tx
+
+  def values: Iterable[SidechainTypes#SCAT] = all.values.map(_.tx)
```
[the remark this answers is not in the document]
Comment (saratnt) [reply sdk/src/main/scala/com/horizen/account/mempool/TxCache.scala:63]: I have the same doubt but I have tried with a map containing 100000 elements and calling values on it takes around 40 ms. It doesn't seem much

File: sdk/src/test/scala/com/horizen/account/mempool/TxCacheTest.scala [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+import com.horizen.account.fixtures.EthereumTransactionFixture
+import org.junit.Assert.{assertEquals, assertFalse, assertTrue}
+import org.junit.Test
+import org.scalatestplus.junit.JUnitSuite
+import org.scalatestplus.mockito.MockitoSugar
+import sparkz.util.ModifierId
+
+import java.math.BigInteger
+
+class TxCacheTest
```
[the remark this answers is not in the document]
Comment (saratnt) [reply sdk/src/test/scala/com/horizen/account/mempool/TxCacheTest.scala:13]: Done

File: sdk/src/test/scala/com/horizen/account/performance/AccountSidechainNodeViewHolderPerfTest.scala [hunk 15]
```diff
@@ -251,8 +256,10 @@ class AccountSidechainNodeViewHolderPerfTest
       out.write(s"Date and time of the test: ${cal.getTime}\n\n")
 
 
-      val numOfTxs = 10016 //this weird number is just to have an integer number of accounts
-      val numOfTxsPerSpammerAccounts = 16
+      //val numOfTxs = 10016 //this weird number is just to have an integer number of accounts
+      val numOfTxs = 100000 //this weird number is just to have an integer number of accounts
+ //     val numOfTxsPerSpammerAccounts = 16
+      val numOfTxsPerSpammerAccounts = 1
```
[the remark this answers is not in the document]
Comment (saratnt) [reply sdk/src/test/scala/com/horizen/account/performance/AccountSidechainNodeViewHolderPerfTest.scala:263]: Done

File: qa/sc_evm_mempool_size.py [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+    assert_equal, \
+    assert_true
+from httpCalls.transaction.allTransactions import allTransactions
+from test_framework.util import forward_transfer_to_sidechain, assert_false
+
+"""
+Check mem pool behaviour when it reached its maximum size.
+
+Configuration:
+    - 1 SC node
```
[the remark this answers is not in the document]
Comment (saratnt) [reply qa/sc_evm_mempool_size.py:18]: I think it is a good idea, remote and local txs may be treated differently and it is worth checking that this is not the case. Done

File: sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala [hunk 17]
```diff
@@ -311,7 +346,8 @@ class MempoolMap(
             (balance.compareTo(tx.maxCost) >= 0) &&
             (currAccountSlots + txSizeInSlots <= MaxSlotsPerAccount)
           ) {
-            all.put(tx.id, tx)
+            //txCache.add(tx)
```
[the remark this answers is not in the document]
Comment (saratnt) [reply sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala:349]: Done

**Review by alsala — CHANGES_REQUESTED** [review]

File: sdk/src/main/scala/com/horizen/account/mempool/TxCache.scala [hunk 18]
```diff
@@ -0,0 +1,78 @@
+package com.horizen.account.mempool
+
+import com.horizen.SidechainTypes
+import com.horizen.account.mempool.MempoolMap.txSizeInSlot
+import sparkz.util.ModifierId
+
+import scala.collection.concurrent.TrieMap
+
+/*
+Thsi class contains all the transactions accepted in the mempool.
```
[the remark this answers is not in the document]
Comment (saratnt) [reply sdk/src/main/scala/com/horizen/account/mempool/TxCache.scala:10]: Done

File: sdk/src/main/scala/com/horizen/SidechainSettings.scala [hunk 19]
```diff
@@ -91,9 +91,11 @@ case class EthServiceSettings(
 
 
 case class AccountMempoolSettings(maxNonceGap: Int = 16,
-                                  maxAccountSlots: Int = 16){
+                                  maxAccountSlots: Int = 16,
+                                  maxMemPoolSlots: Int = 6144){
   require(maxNonceGap > 0, s"Maximum Nonce Gap not positive: $maxNonceGap")
   require(maxAccountSlots > 0, s"Maximum Account Slots not positive: $maxAccountSlots")
+  require(maxMemPoolSlots > 0, s"Maximum Memory Pool Slots not positive: $maxMemPoolSlots")
```
[the remark this answers is not in the document]
Comment (saratnt) [reply sdk/src/main/scala/com/horizen/SidechainSettings.scala:99]: Done

File: sdk/src/main/scala/com/horizen/account/mempool/TxCache.scala (hunk 13 above)
[the remark this answers is not in the document]
Comment (i-Alex) [reply sdk/src/main/scala/com/horizen/account/mempool/TxCache.scala:64]: good

**Review by i-Alex — CHANGES_REQUESTED** [review]
Minor thing left.
Looks good!

**Review by alsala — APPROVED** [review]

**Review by i-Alex — APPROVED** [review]

**Pull request merged**